File: src/main/resources/static/csharp/descending/BubbleSortProject/BubbleSortDescending.cs
using System;
using System.IO;
using System.Diagnostics;

static class Program
{
    static void Main()
    {
        Console.Write("Enter the size of the array: ");
        int n = int.Parse(Console.ReadLine());

        int[] arr = new int[n];
        Random random = new Random();
        for (int i = 0; i < n; i++)
        {
            arr[i] = random.Next(1, 101);
        }

        string originalFile = "original_array.csv";
        string sortedFile = "sorted_result.csv";

        // Write original array to file
        using (StreamWriter writer = new StreamWriter(originalFile))
        {
            for (int i = 0; i < arr.Length; i++)
            {
                writer.Write(arr[i]);
                if (i < arr.Length - 1)
                {
                    writer.Write(",");
                }
            }
            writer.WriteLine();
        }

        // Start timing
        Stopwatch stopwatch = Stopwatch.StartNew();

        BubbleSort(arr);

        // Stop timing
        stopwatch.Stop();
        double totalTime = stopwatch.Elapsed.TotalSeconds;

        // Write sorted array to file
        using (StreamWriter writer = new StreamWriter(sortedFile))
        {
            for (int i = 0; i < arr.Length; i++)
            {
                writer.Write(arr[i]);
                if (i < arr.Length - 1)
                {
                    writer.Write(",");
                }
            }
            writer.WriteLine();
        }

        Console.WriteLine("Original and sorted arrays have been saved to CSV files.");
        Console.WriteLine($"Original array file: {originalFile}");
        Console.WriteLine($"Sorted array file: {sortedFile}");
        Console.WriteLine($"Time taken: {totalTime} seconds.");
    }

    // Bubble sort function
    static void BubbleSort(int[] arr)
    {
        for (int i = 0; i < arr.Length - 1; i++)
        {
            for (int j = 0; j < arr.Length - i - 1; j++)
            {
                if (arr[j] < arr[j + 1])
                {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }
    }
}

File: src/main/resources/static/c/descending/BubbleSortDescending.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

// Function to perform Bubble Sort
void bubbleSort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] < arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

// Function to write an array to a CSV file
void writeToFile(const char* filename, int arr[], int n) {
    FILE *file = fopen(filename, "w");
    if (file == NULL) {
        printf("Error opening file %s\n", filename);
        return;
    }
    for (int i = 0; i < n; i++) {
        fprintf(file, "%d", arr[i]);
        if (i < n - 1) {
            fprintf(file, ",");
        }
    }
    fprintf(file, "\n");
    fclose(file);
}

// Function to read an array from a CSV file
int* readFromFile(const char* filename, int* n) {
    FILE *file = fopen(filename, "r");
    if (file == NULL) {
        printf("Error opening file %s\n", filename);
        return NULL;
    }

    int *arr = NULL;
    int count = 0, value;
    char buffer[4096];

    if (fgets(buffer, sizeof(buffer), file)) {
        char *token = strtok(buffer, ",");
        while (token != NULL) {
            count++;
            arr = realloc(arr, count * sizeof(int));
            if (arr == NULL) {
                printf("Memory allocation failed\n");
                fclose(file);
                return NULL;
            }
            sscanf(token, "%d", &value);
            arr[count - 1] = value;
            token = strtok(NULL, ",");
        }
    }

    *n = count;
    fclose(file);
    return arr;
}

int main() {
    int n;
    printf("Enter the size of the array: ");
    scanf("%d", &n);

    // Define file paths
    const char* originalFile = "../../../static/original_array.csv";
    const char* sortedFile = "../../../static/sorted_result.csv";

    int *arr = NULL;

    // Check if the original array file exists
    FILE *file = fopen(originalFile, "r");
    if (file) {
        // printf("Original array file found. Reading from file...\n");
        fclose(file);
        arr = readFromFile(originalFile, &n);
    } else {
        // printf("Original array file not found. Creating new array...\n");
        arr = (int*)malloc(n * sizeof(int));
        if (arr == NULL) {
            printf("Memory allocation failed\n");
            return 1;
        }

        srand(time(0));
        for (int i = 0; i < n; i++) {
            arr[i] = rand() % 100 + 1;
        }
        writeToFile(originalFile, arr, n);
    }

    clock_t startTime = clock();

    // Sort the array using Bubble Sort
    bubbleSort(arr, n);

    // Write the sorted array to file
    writeToFile(sortedFile, arr, n);

    clock_t endTime = clock();
    double totalTime = (double)(endTime - startTime) / CLOCKS_PER_SEC;

    // Output results with file paths
    // printf("Original array file: %s\n", originalFile);
    // printf("Sorted array file: %s\n", sortedFile);
    printf("Time taken: %.6f seconds.\n", totalTime);

    free(arr);
    return 0;
}
